### PR TITLE
Use PyJWT instead of python-jose

### PR DESCRIPTION
**Pull request by Natim**
Alternate to https://github.com/intuit/oauth-pythonclient/pull/48/

## Context
 
This package indirectly uses python-jose, which is affected by: [GHSA-cjwg-qfpm-7377](https://github.com/advisories/GHSA-cjwg-qfpm-7377) which additionally seems to be [abandoned](https://github.com/mpdavis/python-jose/pull/352#issuecomment-2100580833) by it's maintainers.

Move this package to use [PyJWT](https://github.com/jpadilla/pyjwt/) to generate the JWK instead.

### CODE REVIEW

**Comment by Natim** [comment]
image

**Review by yahel2410 — APPROVED** [review]
LGTM

**Comment by Natim** [comment]
@yahel2410 what's next? Can we merge?

**Comment by robert-mings** [comment]
Hi @Natim,

Thanks for this contribution! The move away from python-jose is a much needed change.

From my testing, I had a few questions/asks:
- It looks like the cryptography backend was being supplied as a `python-jose` dependency. I needed to install the [cryptography package](https://cryptography.io/en/latest/) directly. Was this the same in your testing? If so, we'll need to add this as a dependency.
- [Line 169](https://github.com/intuit/oauth-pythonclient/pull/49/files#diff-cde246b26833c4c8d5c6aff45d668cdc5ac1bb308b6ec56c6aca9a63334ac873R169) - I believe the `verify` method will need to be changed as it was previously supplied by `python-jose`.
- Do you mind adding versioning for [PyJWT](https://github.com/intuit/oauth-pythonclient/pull/49/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552R10)?

**Comment by Natim** [comment]
Hello Robert, no need for the cryptography package anymore with PyJWT.

As for the verify method, PyJWT provides the same method/API. The code is tested and we have been using it in production for more than a month.

In the Python dependencies philosophy, libraries should not enforce dependencies versioning so that project can handle it without library version conflicts.

This is because you can't use two different version of a lib in the same project.

**Comment by Natim** [comment]
> Do you mind adding versioning for [PyJWT](https://github.com/intuit/oauth-pythonclient/pull/49/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552R10)?

Done, thanks

**Comment by robert-mings** [comment]
Thanks @Natim. Great to hear you've been using this in production.

- RE: cryptography - [Based on the PyJWT docs](https://pyjwt.readthedocs.io/en/stable/installation.html#cryptographic-dependencies-optional), the RSA algorithm isn't supported by default until you install the cryptography package. Is there any chance this was being supplied as a dependency from another package in your build? In an isolated environment, I'm receiving exceptions until this is explicitly installed.
- RE: verify method - Can you shared more information on this? I'm not seeing anywhere in the docs where it talks about a verify method/API.
- RE: package versions - Lower bounds look good, thanks for adding this.

**Comment by Natim** [comment]
> cryptography - [Based on the PyJWT docs](https://pyjwt.readthedocs.io/en/stable/installation.html#cryptographic-dependencies-optional), the RSA algorithm isn't supported by default until you install the cryptography package. Is there any chance this was being supplied as a dependency from another package in your build? In an isolated environment, I'm receiving exceptions until this is explicitly installed.

Ok then, let's add it :pray:

**Comment by Natim** [comment]
> verify method - Can you shared more information on this? I'm not seeing anywhere in the docs where it talks about a verify method/API.

It's part of the PyJWK Algorithm API: https://github.com/jpadilla/pyjwt/blob/527fec277e8215a197f8facd3778b359043704ef/jwt/algorithms.py#L180-L185

**Review by jordan-chalupka — APPROVED** [review]

**Comment by yahel2410** [comment]
Can you please proceed and merge this?

**Comment by keaton185** [comment]
@robert-mings are we any closer to being able to merge this one?

**Comment by robert-mings** [comment]
@Natim, @keaton185, @yahel2410 - We're close but I have yet to get this working as it's currently written. Facing `AttributeError: 'PyJWK' object has no attribute 'verify'`. I don't believe it's a namespace issue either as `jwt` is uninstalled.

Can I confirm if others are using this successfully as is?

**Comment by Natim** [comment]
Can you write a test for it I can look at?
It supposed to be an Algorithm object not a JWK object that you use to
verify.

File: intuitlib/utils.py [hunk 1]
```diff
@@ -178,15 +176,14 @@ def get_jwk(kid, jwk_uri):
...
+    :return: Algorithm with the key loaded.
     """
 
     response = requests.get(jwk_uri)
     if response.status_code != 200:
         raise AuthClientError(response)
     data = response.json()
-    keys = next(key for key in data["keys"] if key['kid'] == kid)
-    return keys
+    return PyJWKSet.from_dict(data)[kid]
```
Comment (Natim) [inline intuitlib/utils.py:186]: `PyJWKSet[kid]` returns an Algorithm object: https://github.com/jpadilla/pyjwt/blob/master/jwt/api_jwk.py#L131

File: intuitlib/utils.py [hunk 2]
```diff
@@ -165,9 +164,8 @@ def validate_id_token(id_token, client_id, intuit_issuer, jwk_uri):
         return False
 
     message = id_token_parts[0] + '.' + id_token_parts[1]
-    keys_dict = get_jwk(id_token_header['kid'], jwk_uri)
+    public_key = get_jwk(id_token_header['kid'], jwk_uri)
```
Comment (Natim) [inline intuitlib/utils.py:167]:
```suggestion
    public_key = get_jwk(id_token_header['kid'], jwk_uri).key
```

File: intuitlib/utils.py [hunk 3]
```diff
@@ -165,9 +164,8 @@ def validate_id_token(id_token, client_id, intuit_issuer, jwk_uri):
         return False
 
     message = id_token_parts[0] + '.' + id_token_parts[1]
-    keys_dict = get_jwk(id_token_header['kid'], jwk_uri)
+    public_key = get_jwk(id_token_header['kid'], jwk_uri).key
```
Comment (Natim) [inline intuitlib/utils.py:167]: `PyJWK.key` returns an algorithm

**Comment by Natim** [comment]
@robert-mings can you try again, you were right. I believe we are missing a test for this part of the lib, I was expecting it to exist which I was wrong. The good news is that our production code doesn't seem to use this part of the code.

**Comment by robert-mings** [comment]
Thanks @Natim for adding this. Just an update from my side - this solved the `AttributeError`, however, this introduced a a number of additional errors.

Looks like we needed two additional arguments - padding and algorithm. Easy enough, passing those solves that error:
`is_signature_valid = public_key.verify(message.encode('utf-8'), id_token_signature, algorithm='RS256', padding='PKCS1v15')`

That then runs into another error: `TypeError: Expected instance of hashes.HashAlgorithm`

I'm almost thinking we can simplify and use the `decode` method since this already incorporates logic for validating the signature. This would also allow a bit of cleanup. Maybe something like:

```
   try:
        public_key = get_jwk(id_token_header['kid'], jwk_uri).key
        jwt.decode(id_token, public_key, audience=client_id, algorithms=['RS256'])
        return True
    except:
        return False
```

Thoughts?

**Comment by Natim** [comment]
@robert-mings can you try again like that?

File: intuitlib/utils.py [hunk 4]
```diff
@@ -16,20 +16,19 @@
...
-from jose import jwk
 import requests
-from requests.sessions import Session
 import six
-from requests_oauthlib import OAuth1
-
+import string
+from base64 import b64encode, b64decode, urlsafe_b64decode
+from datetime import datetime
+from jwt import PyJWKSet, PyJWTError
```
Comment (robert-mings) [inline intuitlib/utils.py:25]: Since we are using the `decode` api, let's `import jwt`.

File: intuitlib/utils.py [hunk 5]
```diff
@@ -165,11 +164,12 @@ def validate_id_token(id_token, client_id, intuit_issuer, jwk_uri):
...
-    keys_dict = get_jwk(id_token_header['kid'], jwk_uri)
-
-    public_key = jwk.construct(keys_dict)
-    is_signature_valid = public_key.verify(message.encode('utf-8'), id_token_signature)
-    return is_signature_valid
+    public_key = get_jwk(id_token_header['kid'], jwk_uri).key
+    try:
+        jwt.decode(id_token, public_key, audience=client_id, algorithms=['RS256'])
+        return True
+    except PyJWTError:
```
Comment (robert-mings) [inline intuitlib/utils.py:171]: `jwt.PyJWTError`

File: intuitlib/utils.py [hunk 6]
```diff
@@ -165,11 +164,12 @@ def validate_id_token(id_token, client_id, intuit_issuer, jwk_uri):
         return False
 
     message = id_token_parts[0] + '.' + id_token_parts[1]
```
Comment (robert-mings) [inline intuitlib/utils.py:166]: Can remove the following as they are no longer used:

- `id_token_signature = urlsafe_b64decode(((_correct_padding(id_token_parts[2])).encode('ascii')))`
- `message = id_token_parts[0] + '.' + id_token_parts[1]`

File: intuitlib/utils.py [hunk 7]
```diff
@@ -178,15 +178,14 @@ def get_jwk(kid, jwk_uri):
...
+    :return: Algorithm with the key loaded.
     """
 
     response = requests.get(jwk_uri)
     if response.status_code != 200:
         raise AuthClientError(response)
     data = response.json()
-    keys = next(key for key in data["keys"] if key['kid'] == kid)
-    return keys
+    return PyJWKSet.from_dict(data)[kid]
```
Comment (robert-mings) [inline intuitlib/utils.py:188]: `jwt.PyJWKSet.from_dict(data)[kid]`

**Review by robert-mings — CHANGES_REQUESTED** [review]
Awesome @Natim - this looks good. Only required change is to `import jwt` since we are using the decode api directly (as well as the 2 required references).

I added a couple lines of cleanup as well. Let me know if you're able to make these changes, otherwise I can get this merged and follow-up directly after.

**Review by robert-mings — APPROVED** [review]
Looks great. Thanks again for the contribution @Natim.

Merging now and will get a new version released.

**Pull request merged**